src/lib/core/GroupedCallbackList.h

```diff
@@ -0,0 +1,231 @@
+/*
+ *
+ *    Copyright (c) 2024 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <lib/core/CHIPCallback.h>
+#include <lib/support/CodeUtils.h>
+
+#include <utility>
+
+namespace chip {
+namespace Callback {
+
+namespace detail {
+template <size_t Index>
+void TaggedDequeueGroup(Cancelable * cancelable);
+void EnqueueWithGroup(Cancelable * cancelable, Cancelable *& group, Cancelable * anchor, void (*cancelFn)(Cancelable *));
+void LinkGroup(Cancelable * prev, Cancelable * next);
+inline Cancelable * ClearCancelable(Cancelable * cancelable);
+} // namespace detail
+
+/**
+ * A GroupedCallbackList is similar to a CallbackDeque, but manages groups of
```

Comment:
When reading documentation here, I do not actually yet know what a CallbackDequeue is. Please document it fully on why we use it and how.

Reply:
`CallbackDeque` is the list class defined directly in `CHIPCallback.h` that this class is effectively replacing for certain use cases. I can remove the reference if it is more confusing than useful.

Reply:
I've reworded things to avoid referring to `CallbackDeque` and updated the long comment block and ASCII diagram to explain the anchor element.